client/engine/store/persiststore.go

```diff
@@ -0,0 +1,446 @@
+	}
+}
+func NewPersistStore(key []byte, folder string) Store {
+	ms := PersistStore{}
+	err := os.MkdirAll(folder, os.ModePerm)
+	ms.checkError(err)
+
+	ms.key = common.Bytes2Hex(key)
+	ms.address = crypto.GetAddressFromSecretKeyBytes(key).String()
+
+	ms.objectives, err = buntdb.Open(fmt.Sprintf("%s/objectives_%s.db", folder, ms.address[2:7]))
+	ms.checkError(err)
+
+	ms.channels, err = buntdb.Open(fmt.Sprintf("%s/channels_%s.db", folder, ms.address[2:7]))
+	ms.checkError(err)
+
+	ms.consensusChannels, err = buntdb.Open(fmt.Sprintf("%s/con_channels_%s.db", folder, ms.address[2:7]))
+	ms.checkError(err)
+
+	ms.channelToObjective, err = buntdb.Open(fmt.Sprintf("%s/chan_to_obj_%s.db", folder, ms.address[2:7]))
+	ms.checkError(err)
+	return &ms
+}
+
+func (ms *PersistStore) GetAddress() *types.Address {
+	address := common.HexToAddress(ms.address)
+	return &address
+}
+
+func (ms *PersistStore) GetChannelSecretKey() *[]byte {
+	val := common.Hex2Bytes(ms.key)
+	return &val
+}
+
+func (ms *PersistStore) GetObjectiveById(id protocols.ObjectiveId) (protocols.Objective, error) {
+
+	var obj protocols.Objective
+	err := ms.objectives.View(func(tx *buntdb.Tx) error {
+		objJSON, err := tx.Get(string(id))
+		if err != nil {
+			return err
+		}
+
+		obj, err = decodeObjective(id, []byte(objJSON))
+		if err != nil {
+			return fmt.Errorf("error decoding objective %s: %w", id, err)
+		}
+
+		err = ms.populateChannelData(obj)
+		if err != nil {
+			// return existing objective data along with error
+			return fmt.Errorf("error populating channel data for objective %s: %w", id, err)
+		}
+		return nil
+
+	})
+	if err != nil && errors.Is(err, buntdb.ErrNotFound) {
+		return nil, ErrNoSuchObjective
+	}
+
+	return obj, nil
+}
+
+func (ms *PersistStore) SetObjective(obj protocols.Objective) error {
+	// todo: locking
+	objJSON, err := obj.MarshalJSON()
+
+	if err != nil {
+		return fmt.Errorf("error setting objective %s: %w", obj.Id(), err)
+	}
+
+	err = ms.objectives.Update(func(tx *buntdb.Tx) error {
+
+		_, _, err := tx.Set(string(obj.Id()), string(objJSON), nil)
+		return err
+	})
+
+	if err != nil {
+		return err
+	}
+	for _, rel := range obj.Related() {
+		switch ch := rel.(type) {
+		case *channel.Channel:
+			err := ms.SetChannel(ch)
+			if err != nil {
+				return fmt.Errorf("error setting channel %s from objective %s: %w", ch.Id, obj.Id(), err)
+			}
+		case *consensus_channel.ConsensusChannel:
+			err := ms.SetConsensusChannel(ch)
+			if err != nil {
+				return fmt.Errorf("error setting consensus channel %s from objective %s: %w", ch.Id, obj.Id(), err)
+			}
+		default:
+			return fmt.Errorf("unexpected type: %T", rel)
+		}
+	}
+
+	// Objective ownership can only be transferred if the channel is not owned by another objective
+	var prevOwner protocols.ObjectiveId
+	var isOwned bool = false
+	err = ms.channelToObjective.View(func(tx *buntdb.Tx) error {
+		res, err := tx.Get(string(obj.OwnsChannel().String()))
+		if err != nil {
+			return nil
+		}
```

Comment:
On first glance this looks like a mistake. I think it would improve readability (and safety), to do 
```go
		if err == buntdb.ErrNotFound {
			return nil
		}
```

Reply:
Generally this section is much harder to parse than the corresponding section of the `MemStore`. Some of that difficulty may be unavoidable... but still I wonder if it could be tidied up some more?